### ISSUE TITLE
Improve select from map performance with Mapbox and Google Maps

### ISSUE THREAD

**Issue by lognaturel** [getodk/collect#5087]
#### Software and hardware versions 
Collect v2022.2 beta

#### Problem description
Select one from map is prohibitively slow on Mapbox and Google Maps. A 1000-point form is almost unusable with Mapbox but fine with Google Maps. A 5000-point form is unusuable with both. It loads nearly instantly with osmdroid.

#### Steps to reproduce the problem
Use the thousand-point form [from the forum](https://forum.getodk.org/t/odk-collect-v2022-2-beta/36913/6). Add points using the script in the zip there or use [gilpin.zip](https://github.com/getodk/collect/files/8433510/gilpin.zip)) for 5000.

**Comment by seadowg** [comment]
Yeah I'm seeing the same thing (just testing with Google Maps right now). So something I've noticed:

* Performance appears to be fine when zooming in (to even just a full continent level)
* There seems to be a lag when clicking on items at 5000 points - it looks like `GoogleMapFragment` does a linear search through all the markers on a click so that's probably the problem
* Along with not performing, the 5000 (or even 1000) marker map is kind of useless when zoomed out given how densely the markers end up packed on the screen

So I'd like to suggest we do a few things for Google Maps here:

1. Prevent us doing O(n) work when clicking on a marker (in `GoogleMapFragment#onMarkerClick`)
2. Add clustering to `GoogleMapFragment` using [Google's utility](https://developers.google.com/maps/documentation/android-sdk/utility/marker-clustering) to improve performance (and usability) when the map is zoomed out to show 1000s of markers

An alternative to two is to try and start the map more zoomed in but the performance problem when zoomed out would still exist. I also think clustering will make it easier to read the map at zoomed out levels when you have a lot of markers. We can hopefully make both these changes in the `GoogleMapFragment` (without needing to alter anything in the selection map code) and then investigate similar optimizations for Mapbox. 

Thoughts?

**Comment by seadowg** [comment]
Ok it looks like we can do a similar thing with Mapbox (v9): https://docs.mapbox.com/android/legacy/maps/examples/circle-layer-clustering/

**Comment by lognaturel** [comment]
Thanks for digging in! Glad to hear there are some answers that are not too onerous.

>try and start the map more zoomed in

This intersects with an observation I shared in Slack -- osm seems to launch with the device location centered already. There are so many workflows in which that would be the most useful map centering. Maybe we should just have that behavior for all basemaps for now?

The clustering does sound good, though, so happy for that to be in v2022.2 if that's reasonable.

**Comment by lognaturel** [comment]
@grzesiek2010 you might actually be the best person to look into this because I'm pretty sure it intersects with https://github.com/getodk/collect/pull/5010/. I did a quick profile run with Google Maps:
<img width="1319" alt="Screen Shot 2022-04-10 at 12 06 14 PM" src="https://user-images.githubusercontent.com/967540/162636486-0fb08cb2-df52-4c2e-9a08-4a73eaa09cae.png">

I confirmed by adding some prints of system time that `getBitmapDescriptor(R.drawable.ic_map_point)` takes up to a ms which adds up. And it looks like we call it twice per marker?

On my moto G5, it was never opening because I was getting `Waiting for a blocking GC Alloc`. I didn't have a chance to dig into why this uses so much mem yet. I believe the mem issue is separate from the CPU issue. On another device with more RAM the task completes but still takes several seconds.

> try and start the map more zoomed in
I tried a quick fix for #5093 and that doesn't make a difference. That makes sense, my quick look at `SelectionMapFragment` suggests we're building all the markers no matter what.

I'm having trouble profiling with Mapbox -- Android Studio always crashes for some reason. 

It feels to me like at a high level we're using the mapping engines in a not so ideal way. We'd like to hand off locations and styling info for the engines to lazily load as the markers come into view but instead I think we're doing a bunch of upfront work to do a lot of rendering.

I’m in fact surprised osmdroid is so performant. Maybe because the abstraction layer for mapping engines is closest to its approach to rendering? One thing I hope to do is trace how map markers are initialized there to see how it is performant.

**Comment by grzesiek2010** [comment]
> Select one from map is prohibitively slow on Mapbox and Google Maps. A 1000-point form is almost unusable with Mapbox but fine with Google Maps. A 5000-point form is unusuable with both. It loads nearly instantly with osmdroid.

I'm not sure so I need to ask... are you describing the process of loading data when the progress dialog is displayed or the performance when the data is loaded and you try to play with the items?

> I confirmed by adding some prints of system time that getBitmapDescriptor(R.drawable.ic_map_point) takes up to a ms which adds up. And it looks like we call it twice per marker?

Yes we do and this is something we should change. The problem is that loading x elements we first create markers for every single element with a default `red circle` image and then we update that marker with an appropriate icon see: https://github.com/getodk/collect/blob/master/geo/src/main/java/org/odk/collect/geo/selection/SelectionMapFragment.kt#L353

I tested setting icons only once and the result for all 3 map engines is like 50% faster loading (for 1k items):

| Attempt | Mapbox | Google | Osmdroid |
| :---: | :---: | :---: | :---: |
| master | 17s | 4s | 0.5s | 
| with the fix | 9s | 2s | 0.2s | 

so this is definitely something we need to improve. But this is only about loading, performance once the items are loaded is related to something else.

When it comes to:

> I'm pretty sure it intersects with https://github.com/getodk/collect/pull/5010.

I took a quick look and that fix has impact on memory usage rather than performance itself. Nevertheless it's something we should use in Mapbox too.


It looks like in both Mapbox and Google creating and adding a marker is the step which takes the most time (at least during loading) so probably there is no other option than using clustering as @seadowg has mentioned above. But this is something we would need to test because without it we can't be sure it's going to improve performance.

**Comment by lognaturel** [comment]
> are you describing the process of loading data when the progress dialog is displayed

Yes, this is what I've been focused on because on the moto G5 I can't even get that step to complete!

> setting icons only once

Ahhhh, yeah, we should definitely fix that.

> it's something we should use in Mapbox too

If it's a quick thing for you to do that would be a good first step as part of this. Otherwise maybe file a separate issue?

> creating and adding a marker is the step which takes the most time

Thanks for confirming! I really feel like we're not doing this the way we should be. I'll try to review some docs around that.

**Comment by lognaturel** [comment]
#5100 takes care of setting icons only once and will go out in v2022.2.1. This should halve load time.

We've decided to hold off on further improvements until we get more user feedback and upgrade Mapbox. #5095 may be useful reference when we come back to this.

**Comment by lognaturel** [comment]
We’ll wait until users share a usecase.
